python/lsst/daf/butler/core/composites.py

```diff
@@ -0,0 +1,300 @@
+    storageClass : `StorageClass`
+        `StorageClass` describing the entity to be created from the
+        components.
+    components : `dict`
+        Collection of components from which to assemble a new composite
+        object. Keys correspond to composite names in the `StorageClass`.
+    pytype : `class`, optional
+        Override the type from the `storageClass` to use when assembling
+        the final object.
+
+    Returns
+    -------
+    composite : `object`
+        New composite object assembled from components.
+
+    Raises
+    ------
+    ValueError
+        Some components could not be used to create the object.
+    """
+    if pytype is not None:
+        cls = pytype
+    else:
+        cls = storageClass.pytype
+
+    # Should we check that the storage class components match or are a superset
+    # of the items described in the supplied components?
+
+    # First try to create an instance directly using keyword args
+    try:
+        obj = cls(**components)
+    except TypeError:
+        obj = None
+
+    # Now try to use setters if direct instantiation didn't work
+    if not obj:
+        obj = cls()
+
+        failed = []
+        for name, component in components.items():
+            if component is None:
+                continue
+            for attr in _attrNames(name, getter=False):
+                if hasattr(obj, attr):
+                    if attr == name:  # Real attribute
+                        setattr(obj, attr, component)
+                    else:
+                        setter = getattr(obj, attr)
+                        setter(component)
+                    break
+            else:
+                failed.append(name)
+
+        if failed:
+            raise ValueError("Unhandled components during assembly ({})".format(failed))
+
+    return obj
+
+
+def validComponents(composite, storageClass):
+    """Extract requested components from a composite.
+
+    Parameters
+    ----------
+    composite : `object`
+        Composite from which to extract components.
+    storageClass : `StorageClass`
+        `StorageClass` associated with this composite object.
+        If None, it is assumed this is not to be treated as a composite.
+
+    Returns
+    -------
+    comps : `dict`
+        Non-None components extracted from the composite, indexed by the component
+        name as derived from the `StorageClass`.
+    """
+    components = {}
+    if storageClass is not None and storageClass.components:
+        for c in storageClass.components:
+            if isinstance(composite, collections.Mapping):
+                comp = composite[c]
+            else:
+                try:
+                    comp = genericGetter(composite, c)
+                except AttributeError:
+                    pass
+                else:
+                    if comp is not None:
+                        components[c] = comp
+    return components
+
+
+def hasComponent(composite, componentName):
```

Comment:
Function is not used at the moment since when I wrote it I later realized that in my usage what I wanted to know was whether the component was supported and not None. There is also confusion over attribute versus getter method.